pkg/streamingpromql/operator/binary_operation.go

```diff
@@ -0,0 +1,651 @@
+	for hash, outputSeries := range outputSeriesMap {
+		if len(outputSeries.leftSeriesIndices) == 0 || len(outputSeries.rightSeriesIndices) == 0 {
+			// FIXME: this is incorrect for 'or' and 'unless'
+			// No matching series on at least one side for this output series, so output series will have no samples. Remove it.
+			delete(outputSeriesMap, hash)
+		}
+	}
+
+	allMetadata := make([]SeriesMetadata, 0, len(outputSeriesMap))
+	allSeries := make([]*binaryOperationOutputSeries, 0, len(outputSeriesMap))
+	labelsFunc := b.labelsFunc()
+	leftSeriesUsed := GetBoolSlice(len(b.leftMetadata))[:len(b.leftMetadata)]
+	rightSeriesUsed := GetBoolSlice(len(b.rightMetadata))[:len(b.rightMetadata)]
+
+	for _, outputSeries := range outputSeriesMap {
+		firstSeriesLabels := b.leftMetadata[outputSeries.leftSeriesIndices[0]].Labels
+		allMetadata = append(allMetadata, SeriesMetadata{Labels: labelsFunc(firstSeriesLabels)})
+		allSeries = append(allSeries, outputSeries)
+
+		for _, leftSeriesIndex := range outputSeries.leftSeriesIndices {
+			leftSeriesUsed[leftSeriesIndex] = true
+		}
+
+		for _, rightSeriesIndex := range outputSeries.rightSeriesIndices {
+			rightSeriesUsed[rightSeriesIndex] = true
+		}
+	}
+
+	return allMetadata, allSeries, leftSeriesUsed, rightSeriesUsed
+}
+
+// sortSeries sorts metadata and series in place to try to minimise the number of input series we'll need to buffer in memory.
+//
+// This is critical for minimising the memory consumption of this operator: if we choose a poor ordering of series,
+// we'll need to buffer many input series in memory.
+//
+// At present, sortSeries uses a very basic heuristic to guess the best way to sort the output series, but we could make
+// this more sophisticated in the future.
+func (b *BinaryOperation) sortSeries(metadata []SeriesMetadata, series []*binaryOperationOutputSeries) {
+	// For one-to-one matching, we assume that each output series takes one series from each side of the operator.
+	// If this is true, then the best order is the one in which we read from the highest cardinality side in order.
+	// If we do this, then in the worst case, we'll have to buffer the whole of the lower cardinality side.
+	// (Compare this with sorting so that we read the lowest cardinality side in order: in the worst case, we'll have
+	// to buffer the whole of the higher cardinality side.)
+	//
+	// FIXME: this is reasonable for one-to-one matching, but likely not for one-to-many / many-to-one.
+	// For one-to-many / many-to-one, it would likely be best to buffer the side used for multiple output series (the "one" side),
+	// as we'll need to retain these series for multiple output series anyway.
+
+	var sortInterface sort.Interface
+
+	if len(b.leftMetadata) < len(b.rightMetadata) {
+		sortInterface = favourRightSideSorter{metadata, series}
+	} else {
+		sortInterface = favourLeftSideSorter{metadata, series}
+	}
+
+	sort.Sort(sortInterface)
+}
+
+type favourRightSideSorter struct {
```

Comment:
(nit) Could have these structs embed a generic Sorter which has the metadata+series+Len

Reply:
Interesting idea, the duplication here had annoyed me but I couldn't think of a solution... I'll give this a go.

Reply:
Works beautifully, thanks for the suggestion.